invenio_records_rest/serializers/csv.py

```diff
@@ -0,0 +1,121 @@
+# under the terms of the MIT License; see LICENSE file for more details.
+
+"""Marshmallow based DublinCore serializer for records."""
+
+from __future__ import absolute_import, print_function
+
+import csv
+from six import StringIO
+
+from .base import PreprocessorMixin, SerializerMixinInterface
+from .marshmallow import MarshmallowMixin
+
+
+class CSVSerializer(SerializerMixinInterface, MarshmallowMixin,
+                    PreprocessorMixin):
+    """CSV serializer for records.
+
+    Note: This serializer is not suitable for serializing large number of
+    records.
+    """
+
+    def __init__(self, *args, **kwargs):
+        """Initialize CSVSerializer.
+
+        :param csv_excluded_fields: list of complete paths of the fields that
+                                    should be excluded from the final output
+        :param header_separator: separator that should be used when flattening
+                                 nested dictionary keys
+        """
+        self.csv_excluded_fields = kwargs.pop("csv_excluded_fields", [])
+        self.header_separator = kwargs.pop("header_separator", "_")
+        super(CSVSerializer, self).__init__(*args, **kwargs)
+
+    def serialize(self, pid, record, links_factory=None):
+        """Serialize a single record and persistent identifier.
+
+        :param pid: Persistent identifier instance.
+        :param record: Record instance.
+        :param links_factory: Factory function for record links.
+        """
+        record = self.process_dict(
+            self.transform_record(pid, record, links_factory))
+
+        return self._format_csv([record])
+
+    def serialize_search(self, pid_fetcher, search_result, links=None,
+                         item_links_factory=None):
+        """Serialize a search result.
+
+        :param pid_fetcher: Persistent identifier fetcher.
+        :param search_result: Elasticsearch search result.
+        :param links: Dictionary of links to add to response.
+        :param item_links_factory: Factory function for record links.
+        """
+        records = []
+        for hit in search_result['hits']['hits']:
+            processed_hit = self.transform_search_hit(
+                pid_fetcher(hit['_id'], hit['_source']),
+                hit,
+                links_factory=item_links_factory,
+            )
+            records.append(self.process_dict(processed_hit))
+
+        return self._format_csv(records)
+
+    def process_dict(self, dictionary):
+        """Transform record dict with nested keys to a flat dict."""
+        return self._flatten(dictionary)
+
+    def _format_csv(self, records):
+        """Return the list of records as a CSV string."""
+        # build a unique list of all records keys as CSV headers
+        headers = set()
+        for rec in records:
+            headers.update(rec.keys())
+
+        # write the CSV output in memory
+        output = StringIO()
```

Comment:
TBH, I prefer the second solution from [this SO thread](https://stackoverflow.com/a/32609533) (the one with the `Line` class), it looks a bit higher-level (i.e. no `.seek` and `.truncate` calls)

Reply:
You are right, this is much better! Will change that! 👍